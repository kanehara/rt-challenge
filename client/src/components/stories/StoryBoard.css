.storyBoard {
    --smallFontSize: .75em;
    font-size: var(--smallFontSize);
}

.storyBoard .ui.label {
    font-size: var(--smallFontSize);
    cursor: pointer;
}

.storyBoard .ui.card {
    width: initial;
}

.storyBoard .ui.label:hover {
    background-color: #c3c3c3;
}

@media screen and (min-width: 768px) {
    .storyBoard {
        font-size: 1em;
    }

    .storyBoard .ui.label {
        font-size: .85em;
    }
}