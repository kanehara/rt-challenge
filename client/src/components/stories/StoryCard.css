.ui.label.tiny {
    font-size: .45em;
    cursor: pointer;
    margin: 0.18em;
}

.ui.button {
    font-size: var(--smallFontSize);
}

.ui.card {
    width: initial;
}

.ui.label:hover {
    background-color: #c3c3c3;
}

.ui.label.dummyAvatar:hover {
    cursor: default;
    background-color: #e8e8e8;
}

.ui.label.dummyAvatar {
    font-size: var(--smallFontSize);
    margin: .18em;
}

a {
    color: var(--secondaryColor);
}

@media screen and (min-width: 768px) {
    .ui.button {
        font-size: .85em;
    }

    .ui.label.tiny {
        font-size: .55em;
    }

    .ui.label.dummyAvatar {
        font-size: .85em;
    }
}