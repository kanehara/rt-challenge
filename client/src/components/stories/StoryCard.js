import React from 'react'
import 'semantic-ui-label/label.min.css'
import 'semantic-ui-icon/icon.min.css'
import 'semantic-ui-card/card.min.css'
import { withRouter } from 'react-router-dom'



export const StoryCard = ({history, story, match}) => {
    let storyTypeIcon = null
    switch(story.story_type) {
        case 'feature':
            storyTypeIcon = (<i className="code icon"/>)
            break;
        case 'bug':
            storyTypeIcon = (<i className="bug icon"/>)
            break;
        case 'chore':
            storyTypeIcon = (<i className="cubes icon"/>)
            break;
        case 'release':
            storyTypeIcon = (<i className="checkmark box icon"/>)
            break;
        default:
            console.error(`Invalid story type for story: ${story}`)
    }

    const owners = story.owners.map(owner => (
        <div key={owner.poid} className="ui image label small dummyAvatar">
            <img src="/dummyAvatar.jpg" alt="dummy avatar"/>
            {owner.initials}
        </div>
    ))

    const storyType = story.story_type === 'feature'
        ? (<h4>Estimate: {story.estimate ? story.estimate : 'Unestimated'}</h4>)
        : null

    const labels = story.labels.map(label => (
        <div key={label.id} className="ui label tiny"
             onClick={() => history.push(`/projects/${match.params.projectId}/stories/${label.id}`)}>
            {label.name}
        </div>
    ))

    return (
        <div key={story.id} className="ui card">
            <div className="content">
                <div className="header">
                    <a href={story.url} target="_blank">{story.name}</a>
                </div>
            </div>
            <div className="content">
                <div>{owners}</div>
                <h4>{storyTypeIcon} {story.story_type}</h4>
                {storyType}
                <h4>State: {story.current_state}</h4>
                {labels}
            </div>
        </div>
    )
}

export default withRouter(StoryCard)