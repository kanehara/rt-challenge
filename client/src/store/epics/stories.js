import {Observable} from 'rxjs/Observable'
import * as ActionTypes from '../actionTypes';
import axios from './util/axiosClient';

export const getAllStories = action$ => action$
    .ofType(ActionTypes.GET_STORIES)
    .switchMap(({ payload: { projectId, labelId } }) => {
      if (labelId) {
          const stories$ = Observable.fromPromise(axios.get(`/projects/${projectId}/stories?label=${labelId}`))
              .map(res => res.data)
              .catch(error => Observable.of({
                  type: ActionTypes.GET_STORIES_FAILURE,
                  payload: {error},
              }))
          const label$ = Observable.fromPromise(axios.get(`/projects/${projectId}/labels/${labelId}`))
              .map(res => res.data.name)
              .catch(error => Observable.of({
                  type: ActionTypes.GET_LABEL_FAILURE,
                  payload: {error},
              }))
          return Observable.zip(
              stories$,
              label$,
              (stories, label) => ({
                 type: ActionTypes.GET_STORIES_SUCCESS,
                 payload: {stories, label}
              }))
      } else {
        return Observable.fromPromise(axios.get(`/projects/${projectId}/stories`))
            .map(res => res.data)
            .map(stories => ({
                type: ActionTypes.GET_STORIES_SUCCESS,
                payload: {stories},
            }))
            .catch(error => Observable.of({
                type: ActionTypes.GET_STORIES_FAILURE,
                payload: {error},
            }))
      }
    });