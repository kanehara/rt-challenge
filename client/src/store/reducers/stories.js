import * as actionTypes from '../actionTypes'

const initialState = { stories: [], status: 'initial' }

export const stories = (state = initialState, action) => {
  switch(action.type) {
    case actionTypes.GET_STORIES:
      return {
        stories: [],
        status: 'loading'
      };
    case actionTypes.GET_STORIES_SUCCESS:
      return {
        stories: action.payload.stories,
        status: 'success'
      };
    case actionTypes.GET_STORIES_FAILURE:
      return {
        stories: [],
        status: 'failure'
      };
    default:
      return state;
  }
};