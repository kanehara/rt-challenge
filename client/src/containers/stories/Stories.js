import React from 'react'
import {connect} from 'react-redux'
import _ from 'lodash'

import * as actions from '../../store/actions'
import StoryBoard from '../../components/stories/StoryBoard'

const mapStateToProps = (state) => ({
  status: state.stories.status,
  stories: state.stories.stories,
  label: state.stories.label,
  projects: state.projects
})

const mapDispatchToProps = (dispatch) => ({
  getStories: (payload) => dispatch(actions.getStories(payload)),
  getProjects: (payload) => dispatch(actions.getProjects(payload))
})

export class Stories extends React.Component {

  componentWillMount() {
    this.props.getStories(this.props.match.params)
    if (this.props.projects.status === 'initial') {
        this.props.getProjects(this.props.match.params)
    }
  }

  componentWillUpdate(nextProps) {
    if (this.props.match.params.labelId !== nextProps.match.params.labelId) {
      this.props.getStories(nextProps.match.params)
    }
  }

  render() {
    const currentProject = _.head(this.props.projects.projects
        .filter(p => String(p.id) === this.props.match.params.projectId))

    const boardDescription = (
      <div>
        <h2>{currentProject ? currentProject.name : ''}</h2>
        <h4>Showing: {this.props.label ? this.props.label : 'Current Iteration' }</h4>
      </div>
    )

    switch(this.props.status) {
        case 'loading':
            return (
                <div>
                  {boardDescription}
                  <h3>Loading...</h3>
                </div>
            )
        case 'failure':
            return (
                <div>
                    <h3>Failed to retrieve stories</h3>
                </div>
            )
        case 'success':
            return (
                <div>
                  {boardDescription}
                  <StoryBoard stories={this.props.stories} />
                </div>
            )
        default:
            return (<div><h3>500</h3></div>)
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Stories)