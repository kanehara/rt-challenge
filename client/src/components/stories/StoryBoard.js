import React from 'react'
import './StoryBoard.css'
import StoryCard from './StoryCard'

export const buildStoriesByStatus = stories => {
  return stories.reduce((built, story) => {
    if (built[story.current_state]) {
      built[story.current_state].push(story)
    }
    return built;
  }, {unscheduled: [], planned: [], unstarted: [], started: [], finished: [], delivered: [], accepted: [], rejected: []});
}

/**
 * Concatenates all open tickets from the storiesByStatus map
 *
 * TODO: add ability to show non current iteration
 */
export const getOpenStories = ({unscheduled, planned, unstarted}) => [].concat(unstarted, planned, unscheduled)

/**
 * Concatenates all in progress tickets from the storiesByStatusMap
 */
export const getInProgressTickets = ({ started, rejected }) => [].concat(started, rejected)

export default ({stories}) => {
  const storiesByStatus = buildStoriesByStatus(stories)
  const inProgressStories = getInProgressTickets(storiesByStatus)
  const finishedStories = storiesByStatus.finished
  const inAcceptanceStories = storiesByStatus.delivered
  const closedStories = storiesByStatus.accepted

  const textAlignCenter = {
    textAlign: 'center'
  }

  return (
    <div className="grid four storyBoard">
      <h2 style={textAlignCenter}>In Progress</h2>
      <h2 style={textAlignCenter}>Finished</h2>
      <h2 style={textAlignCenter}>In Acceptance</h2>
      <h2 style={textAlignCenter}>Closed</h2>
      <div className="inProgress">
        {inProgressStories.map(story => (
            <StoryCard key={story.id} story={story} />
        ))}
      </div>
      <div className="finished">
        {finishedStories.map(story => (
            <StoryCard key={story.id} story={story} />
        ))}
      </div>
      <div className="inAcceptance">
        {inAcceptanceStories.map(story => (
            <StoryCard key={story.id} story={story} />
        ))}
      </div>
      <div className="closed">
        {closedStories.map(story => (
            <StoryCard key={story.id} story={story} />
        ))}
      </div>
    </div>
  )}