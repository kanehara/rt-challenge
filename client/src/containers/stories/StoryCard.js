import React from 'react'
import 'semantic-ui-label/label.min.css'
import 'semantic-ui-icon/icon.min.css'
import 'semantic-ui-card/card.min.css'
import 'semantic-ui-button/button.min.css'
import { withRouter } from 'react-router-dom'

import '../../components/stories/StoryCard.css'
import StoryOwner from '../../components/stories/details/StoryOwner'
import StoryType from '../../components/stories/details/StoryType'
import StoryEstimate from '../../components/stories/details/StoryEstimate'
import StoryLabel from '../../components/stories/details/StoryLabel'

export const StoryCard = ({history, story, match}) => {

    const owners = story.owners.map(owner => (<StoryOwner key={owner.poid} owner={owner}/>))

    const estimate = story.story_type === 'feature'
        ? <StoryEstimate estimate={story.estimate}/>
        : null

    const labelClickHandler = ({id}) => history.push(`/projects/${match.params.projectId}/stories/${id}`)
    const labels = story.labels.map(label => <StoryLabel key={label.id} label={label} clickHandler={labelClickHandler} />)

    return (
        <div key={story.id} className="ui card">
            <div className="content">
                <div className="header">
                    <a href={story.url} target="_blank">{story.name}</a>
                    <div>
                        {labels}
                    </div>
                </div>
            </div>
            <div className="content">
                <div>{owners}</div>
                <StoryType story={story}/>
                {estimate}
                <h4>State: {story.current_state}</h4>
                <div className="ui button">Action</div>
            </div>
        </div>
    )
}

export default withRouter(StoryCard)