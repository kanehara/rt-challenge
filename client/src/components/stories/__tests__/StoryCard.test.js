import React from 'react'
import {shallow} from 'enzyme'
import renderer from 'react-test-renderer'
import {Router} from 'react-router'

import {StoryCard} from '../../../containers/stories/StoryCard'

describe('StoryCard rendering', () => {
    it('renders story details', () => {
        const mockStory = {
            id: 1,
            story_type: 'feature',
            url: 'test.com',
            name: 'test story',
            owners: [
                {
                    poid: 1,
                    initials: 'yk'
                }
            ],
            estimate: 1,
            labels: [
                {
                    id: 1,
                    name: 'test label'
                }
            ],
            current_state: 'finished'
        }

        const mockHistory = {

        }

        const mockMatch = {
            params: {
                projectId: '1'
            }
        }

        const result = renderer.create(
            <StoryCard story={mockStory} history={mockHistory} match={mockMatch}/>
        ).toJSON()
        expect(result).toMatchSnapshot()
    })
})