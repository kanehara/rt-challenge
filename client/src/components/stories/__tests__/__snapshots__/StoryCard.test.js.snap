// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryCard rendering renders story details 1`] = `
<div
  className="ui card"
>
  <div
    className="content"
  >
    <div
      className="header"
    >
      <a
        href="test.com"
        target="_blank"
      >
        test story
      </a>
    </div>
  </div>
  <div
    className="content"
  >
    <div>
      <div
        className="ui image label small dummyAvatar"
      >
        <img
          alt="dummy avatar"
          src="/dummyAvatar.jpg"
        />
        yk
      </div>
    </div>
    <h4>
      <i
        className="code icon"
      />
       
      feature
    </h4>
    <h4>
      Estimate: 
      1
    </h4>
    <h4>
      State: 
      finished
    </h4>
    <div
      className="ui label tiny"
      onClick={[Function]}
    >
      test label
    </div>
  </div>
</div>
`;
